### ISSUE TITLE
New version 0.9.1 gives "Error Probe could not be created" on macOS Catalina

### ISSUE THREAD

**Issue by coolaj86** [probe-rs/cargo-embed#99]
I realize that the error is probably in a dependency, not cargo-embed itself, but I'm not clear on how to drill down further.

# Describe the bug

After upgrading to `cargo-flash 0.9.0` and `cargo-embed 0.9.1` on macOS I can no longer flash my STM32 Blue Pill. Instead I get this error message:

```txt
   Compiling blue_pill_base v0.1.0 (/Users/aj/BluePill_Rust/utah-rust-blue-pill-code)
    Finished release [optimized + debuginfo] target(s) in 5.29s
    Flashing /Users/aj/BluePill_Rust/utah-rust-blue-pill-code/target/thumbv7m-none-eabi/release/blue_pill_base
       Error Probe could not be created

Caused by:
    0: Pipe error
    1: Pipe error
```

## Workaround

Switching back to `cargo-flash 0.8.0` and `cargo-embed 0.8.0` solves this problem.

# To Reproduce

1.  I'm working from these instructions https://cgit.pinealservo.com/BluePill_Rust/resources/issues/1, trying to update them for v0.9.x.
2. I have a Blue Pill STM32 plugged into my USB, with libftdi drivers installed on macOS.
    ```bash
    brew install libftdi
    Updating Homebrew...
    Warning: libftdi 1.5 is already installed and up-to-date
    To reinstall 1.5, run `brew reinstall libftdi`
    ```
3. This is what I did:
    ```bash
    git clone git clone https://cgit.pinealservo.com/BluePill_Rust/blue_pill_base ./utah-rust-blue-pill-code/
    pushd /utah-rust-blue-pill-code/
    rustup run 1.46.0 \
	    cargo install cargo-flash --version 0.9.0
    rustup run 1.46.0 \
	    cargo install cargo-embed --version 0.9.1
    cargo embed --release
    ```

# Expected behavior

I expect that v0.9.x should flash my STM32 as well or better than v0.8.x

# Desktop (please complete the following information):

 - macOS Catalina

# Additional Context

```bash
cargo flash --version
cargo-flash 0.9.0

cargo flash --list-probes
The following devices were found:
[0]: STLink V2 (VID: 0483, PID: 3748, Serial: 7DC299030110134753384C4E00, STLink)
```

<details>
<summary>Compared dependencies between the two versions</summary>

```diff
   Compiling cargo-embed v0.8.0				      |	   Compiling cargo-embed v0.9.1

   Compiling addr2line v0.13.0				      <
   Compiling adler v0.2.3					   Compiling adler v0.2.3
   Compiling adler32 v1.2.0					   Compiling adler32 v1.2.0
   Compiling aho-corasick v0.7.14				   Compiling aho-corasick v0.7.14
   Compiling ansi_term v0.11.0					   Compiling ansi_term v0.11.0
   Compiling anyhow v1.0.33					   Compiling anyhow v1.0.33
   Compiling arc-swap v0.4.7					   Compiling arc-swap v0.4.7
							      >	   Compiling arrayvec v0.5.1
   Compiling async-channel v1.5.1				   Compiling async-channel v1.5.1
   Compiling async-executor v1.3.0				   Compiling async-executor v1.3.0
   Compiling async-global-executor v1.4.0			   Compiling async-global-executor v1.4.0
   Compiling async-io v1.1.10					   Compiling async-io v1.1.10
   Compiling async-mutex v1.4.0					   Compiling async-mutex v1.4.0
   Compiling async-std v1.6.5					   Compiling async-std v1.6.5
   Compiling async-task v4.0.2					   Compiling async-task v4.0.2
   Compiling atomic-waker v1.0.0				   Compiling atomic-waker v1.0.0
   Compiling atty v0.2.14					   Compiling atty v0.2.14
   Compiling autocfg v1.0.1					   Compiling autocfg v1.0.1
   Compiling backtrace v0.3.53				      <
   Compiling base64 v0.12.3					   Compiling base64 v0.12.3
   Compiling bitfield v0.13.2					   Compiling bitfield v0.13.2
   Compiling bitflags v1.2.1					   Compiling bitflags v1.2.1
							      >	   Compiling bitvec v0.17.4
   Compiling blocking v1.0.2					   Compiling blocking v1.0.2
   Compiling byteorder v1.3.4					   Compiling byteorder v1.3.4
   Compiling cache-padded v1.1.1				   Compiling cache-padded v1.1.1
   Compiling cargo-project v0.2.4			      |	   Compiling cargo_metadata v0.11.4
							      >	   Compiling cargo_toml v0.8.1
   Compiling cassowary v0.3.0					   Compiling cassowary v0.3.0
   Compiling cc v1.0.61						   Compiling cc v1.0.61
   Compiling cfg-if v0.1.10					   Compiling cfg-if v0.1.10
   Compiling cfg-if v1.0.0				      <
   Compiling chrono v0.4.19					   Compiling chrono v0.4.19
   Compiling clap v2.33.3					   Compiling clap v2.33.3
   Compiling colored v1.9.3				      |	   Compiling colored v2.0.0
   Compiling concurrent-queue v1.2.2				   Compiling concurrent-queue v1.2.2
   Compiling config v0.9.3				      |	   Compiling config v0.10.1
   Compiling console v0.12.0					   Compiling console v0.12.0
   Compiling crc32fast v1.2.0					   Compiling crc32fast v1.2.0
   Compiling crossbeam-utils v0.7.2				   Compiling crossbeam-utils v0.7.2
   Compiling crossterm v0.17.7					   Compiling crossterm v0.17.7
   Compiling derivative v2.1.1					   Compiling derivative v2.1.1
   Compiling dtoa v0.4.6					   Compiling dtoa v0.4.6
   Compiling either v1.6.1					   Compiling either v1.6.1
   Compiling enum-primitive-derive v0.2.1			   Compiling enum-primitive-derive v0.2.1
   Compiling env_logger v0.7.1					   Compiling env_logger v0.7.1
   Compiling envy v0.4.1				      <
   Compiling event-listener v2.5.1				   Compiling event-listener v2.5.1
   Compiling failure v0.1.8				      <
   Compiling failure_derive v0.1.8			      <
   Compiling fallible-iterator v0.2.0				   Compiling fallible-iterator v0.2.0
   Compiling fastrand v1.4.0					   Compiling fastrand v1.4.0
   Compiling filetime v0.2.12					   Compiling filetime v0.2.12
   Compiling flate2 v1.0.18					   Compiling flate2 v1.0.18
   Compiling futures v0.3.6					   Compiling futures v0.3.6
   Compiling futures-channel v0.3.6				   Compiling futures-channel v0.3.6
   Compiling futures-core v0.3.6				   Compiling futures-core v0.3.6
   Compiling futures-executor v0.3.6				   Compiling futures-executor v0.3.6
   Compiling futures-io v0.3.6					   Compiling futures-io v0.3.6
   Compiling futures-lite v1.11.1				   Compiling futures-lite v1.11.1
   Compiling futures-macro v0.3.6				   Compiling futures-macro v0.3.6
   Compiling futures-sink v0.3.6				   Compiling futures-sink v0.3.6
   Compiling futures-task v0.3.6				   Compiling futures-task v0.3.6
   Compiling futures-util v0.3.6				   Compiling futures-util v0.3.6
   Compiling gdb-protocol v0.1.0				   Compiling gdb-protocol v0.1.0
   Compiling gdb-server v0.8.0				      |	   Compiling gdb-server v0.9.0
   Compiling gimli v0.21.0				      <
   Compiling gimli v0.22.0					   Compiling gimli v0.22.0
   Compiling glob v0.3.0				      <
   Compiling goblin v0.2.3					   Compiling goblin v0.2.3
   Compiling hashbrown v0.9.1					   Compiling hashbrown v0.9.1
   Compiling heck v0.3.1					   Compiling heck v0.3.1
							      >	   Compiling hex v0.4.2
   Compiling hidapi v1.2.3					   Compiling hidapi v1.2.3
   Compiling humantime v1.3.0					   Compiling humantime v1.3.0
   Compiling ihex v3.0.0					   Compiling ihex v3.0.0
   Compiling indexmap v1.6.0					   Compiling indexmap v1.6.0
   Compiling indicatif v0.14.0				      |	   Compiling indicatif v0.15.0
   Compiling itertools v0.9.0				      <
   Compiling itoa v0.4.6					   Compiling itoa v0.4.6
   Compiling jaylink v0.1.5					   Compiling jaylink v0.1.5
   Compiling jep106 v0.2.4					   Compiling jep106 v0.2.4
   Compiling kv-log-macro v1.0.7				   Compiling kv-log-macro v1.0.7
   Compiling lazy_static v1.4.0					   Compiling lazy_static v1.4.0
							      >	   Compiling lexical-core v0.7.4
   Compiling libc v0.2.79					   Compiling libc v0.2.79
   Compiling libflate v1.0.2					   Compiling libflate v1.0.2
   Compiling libflate_lz77 v1.0.0				   Compiling libflate_lz77 v1.0.0
							      >	   Compiling libftdi1-sys v1.0.0-alpha3
   Compiling libusb1-sys v0.4.2					   Compiling libusb1-sys v0.4.2
   Compiling linked-hash-map v0.5.3				   Compiling linked-hash-map v0.5.3
   Compiling lock_api v0.3.4					   Compiling lock_api v0.3.4
   Compiling log v0.4.11					   Compiling log v0.4.11
   Compiling memchr v2.3.3					   Compiling memchr v2.3.3
   Compiling miniz_oxide v0.4.3					   Compiling miniz_oxide v0.4.3
   Compiling mio v0.7.3						   Compiling mio v0.7.3
   Compiling nb-connect v1.0.2					   Compiling nb-connect v1.0.2
   Compiling nom v4.2.3					      |	   Compiling nom v5.1.2
   Compiling num-integer v0.1.43				   Compiling num-integer v0.1.43
   Compiling num-traits v0.2.12					   Compiling num-traits v0.2.12
   Compiling num_cpus v1.13.0					   Compiling num_cpus v1.13.0
   Compiling number_prefix v0.3.0				   Compiling number_prefix v0.3.0
   Compiling object v0.20.0				      <
   Compiling object v0.21.1					   Compiling object v0.21.1
   Compiling once_cell v1.4.1					   Compiling once_cell v1.4.1
   Compiling parking v2.0.0					   Compiling parking v2.0.0
   Compiling parking_lot v0.10.2				   Compiling parking_lot v0.10.2
   Compiling parking_lot_core v0.7.2				   Compiling parking_lot_core v0.7.2
   Compiling pin-project v0.4.27				   Compiling pin-project v0.4.27
   Compiling pin-project-internal v0.4.27			   Compiling pin-project-internal v0.4.27
   Compiling pin-project-lite v0.1.10				   Compiling pin-project-lite v0.1.10
   Compiling pin-utils v0.1.0					   Compiling pin-utils v0.1.0
   Compiling pkg-config v0.3.19					   Compiling pkg-config v0.3.19
   Compiling plain v0.2.3					   Compiling plain v0.2.3
   Compiling polling v2.0.1					   Compiling polling v2.0.1
   Compiling probe-rs v0.8.0				      |	   Compiling probe-rs v0.9.0
   Compiling probe-rs-rtt v0.3.0			      |	   Compiling probe-rs-cli-util v0.9.0
							      >	   Compiling probe-rs-rtt v0.4.0
   Compiling probe-rs-t2rust v0.6.0				   Compiling probe-rs-t2rust v0.6.0
   Compiling proc-macro-error v1.0.4				   Compiling proc-macro-error v1.0.4
   Compiling proc-macro-error-attr v1.0.4			   Compiling proc-macro-error-attr v1.0.4
   Compiling proc-macro-hack v0.5.18				   Compiling proc-macro-hack v0.5.18
   Compiling proc-macro-nested v0.1.6				   Compiling proc-macro-nested v0.1.6
   Compiling proc-macro2 v0.4.30			      <
   Compiling proc-macro2 v1.0.24				   Compiling proc-macro2 v1.0.24
   Compiling quick-error v1.2.3					   Compiling quick-error v1.2.3
   Compiling quote v0.6.13				      <
   Compiling quote v1.0.7					   Compiling quote v1.0.7
   Compiling recap v0.1.1				      |	   Compiling radium v0.3.0
   Compiling recap-derive v0.1.1			      <
   Compiling regex v1.4.1					   Compiling regex v1.4.1
   Compiling regex-syntax v0.6.20				   Compiling regex-syntax v0.6.20
   Compiling rle-decode-fast v1.0.1				   Compiling rle-decode-fast v1.0.1
   Compiling rusb v0.6.4					   Compiling rusb v0.6.4
   Compiling rustc-cfg v0.4.0				      <
   Compiling rustc-demangle v0.1.17			      <
   Compiling ryu v1.0.5						   Compiling ryu v1.0.5
   Compiling scopeguard v1.1.0					   Compiling scopeguard v1.1.0
   Compiling scroll v0.10.2					   Compiling scroll v0.10.2
   Compiling scroll_derive v0.10.3				   Compiling scroll_derive v0.10.3
							      >	   Compiling semver v0.10.0
							      >	   Compiling semver-parser v0.7.0
   Compiling serde v1.0.117					   Compiling serde v1.0.117
   Compiling serde_derive v1.0.117				   Compiling serde_derive v1.0.117
   Compiling serde_json v1.0.59					   Compiling serde_json v1.0.59
   Compiling serde_yaml v0.8.13					   Compiling serde_yaml v0.8.13
   Compiling signal-hook v0.1.16				   Compiling signal-hook v0.1.16
   Compiling signal-hook-registry v1.2.1			   Compiling signal-hook-registry v1.2.1
   Compiling slab v0.4.2					   Compiling slab v0.4.2
   Compiling smallvec v1.4.2					   Compiling smallvec v1.4.2
   Compiling stable_deref_trait v1.2.0				   Compiling stable_deref_trait v1.2.0
							      >	   Compiling static_assertions v1.1.0
   Compiling strsim v0.8.0					   Compiling strsim v0.8.0
   Compiling structopt v0.3.20					   Compiling structopt v0.3.20
   Compiling structopt-derive v0.4.13				   Compiling structopt-derive v0.4.13
   Compiling svg v0.7.2					      |	   Compiling svg v0.8.0
   Compiling syn v0.15.44				      <
   Compiling syn v1.0.45					   Compiling syn v1.0.45
   Compiling synstructure v0.12.4			      <
   Compiling tar v0.4.30					   Compiling tar v0.4.30
   Compiling termcolor v1.1.0					   Compiling termcolor v1.1.0
   Compiling terminal_size v0.1.13				   Compiling terminal_size v0.1.13
   Compiling termios v0.3.3					   Compiling termios v0.3.3
   Compiling textwrap v0.11.0					   Compiling textwrap v0.11.0
							      >	   Compiling textwrap v0.12.1
   Compiling thiserror v1.0.21					   Compiling thiserror v1.0.21
   Compiling thiserror-impl v1.0.21				   Compiling thiserror-impl v1.0.21
   Compiling thread_local v1.0.1				   Compiling thread_local v1.0.1
   Compiling time v0.1.44					   Compiling time v0.1.44
   Compiling toml v0.4.10				      |	   Compiling toml v0.5.7
   Compiling tui v0.9.5					      |	   Compiling tui v0.10.0
   Compiling unicode-segmentation v1.6.0			   Compiling unicode-segmentation v1.6.0
   Compiling unicode-width v0.1.8				   Compiling unicode-width v0.1.8
   Compiling unicode-xid v0.1.0				      <
   Compiling unicode-xid v0.2.1					   Compiling unicode-xid v0.2.1
   Compiling vec-arena v1.0.0					   Compiling vec-arena v1.0.0
   Compiling vec_map v0.8.2					   Compiling vec_map v0.8.2
   Compiling version_check v0.1.5			      <
   Compiling version_check v0.9.2				   Compiling version_check v0.9.2
   Compiling waker-fn v1.1.0					   Compiling waker-fn v1.1.0
   Compiling wasmparser v0.57.0					   Compiling wasmparser v0.57.0
   Compiling xattr v0.2.2					   Compiling xattr v0.2.2
   Compiling yaml-rust v0.4.4					   Compiling yaml-rust v0.4.4
```

</details>

P.S. My guess is that `probe-rs` or `probe-rs-rtt` is actually the culprit.

**Comment by Yatekii** [comment]
Thanks for reporting this! It is a known issue (already reported in the probe-rs repo) but I have no way of debugging it as I do not own a mac.
The error must be within rusb or in the way we use it. Maybe thid is fixed on master. Could you try installing from cargo-embed master with probe-rs master? :)

**Comment by coolaj86** [comment]
@Yatekii I'm still fairly new to the rust tooling ecosystem. How does one specify to install from master?

```bash
git clone https://github.com/probe-rs/cargo-embed
git push ./cargo-embed/
cargo ??? probes-rs --branch master
cargo install --path .
```

**Comment by Yatekii** [comment]
Try `cargo install --git https://github.com/probe-rs/cargo-embed` :)

**Comment by coolaj86** [comment]
```
cargo install --git https://github.com/probe-rs/cargo-embed
cargo embed --release

    Finished release [optimized + debuginfo] target(s) in 0.33s
      Config default
      Target /Users/aj/BluePill_Rust/utah-rust-blue-pill-code/target/thumbv7m-none-eabi/release/blue_pill_base
     Erasing sectors ✔ [00:00:00] [################################################################]   7.00KB/  7.00KB @  24.59KB/s (eta 0s )
 Programming pages   ✔ [00:00:00] [################################################################]   7.00KB/  7.00KB @   6.78KB/s (eta 0s )
    Finished flashing in 0.622s
       Error Error attaching to RTT: Error communicating with probe: An error with the usage of the probe occured
```

**Comment by Yatekii** [comment]
Yay, progress!

I am not sure about the next one, there is parts of the message missing :)

Do you have some more info on what firmware you run?

**Comment by coolaj86** [comment]
> Do you have some more info on what firmware you run?

I'm not clear on what you mean by `firmware` in this case.

I think of `firmware` as what I'm trying to flash to the microcontroller, but I don't think that's what you mean since `cargo embed 0.8.0` is what I've been using to flash the device with said firmware (which is this blinky light program: https://cgit.pinealservo.com/BluePill_Rust/blue_pill_base).

What command might I run to give you more details.

**Comment by coolaj86** [comment]
In other news: `cargo-flash` from master _does_ work:

```bash
cargo install --git https://github.com/probe-rs/cargo-flash
cargo flash --release
```

```txt
    Finished release [optimized + debuginfo] target(s) in 0.03s
    Flashing /Users/aj/BluePill_Rust/utah-rust-blue-pill-code/target/thumbv7m-none-eabi/release/blue_pill_base
     Erasing sectors ✔ [00:00:00] [#######################]   7.00KB/  7.00KB @  31.02KB/s (eta 0s )
 Programming pages   ✔ [00:00:00] [#######################]   7.00KB/  7.00KB @   7.62KB/s (eta 0s )
```

**Comment by Yatekii** [comment]
Well about the firmware question: If the firmware uses `wfi();` which RTIC applications with no idle() for example do, RTT wont work on most chips :) That's why I asked.

But the error does not look firmware related. You could try setting the `speed` property to 100. This will be slow but for experimentation purposes it could help :)

**Comment by coolaj86** [comment]
@Yatekii I updated `Embed.toml`, but it didn't make a difference.

```yml
[default.probe]
# USB vendor ID
# usb_vid = "1337"
# USB product ID
# usb_pid = "1337"
# Serial number
# serial = "12345678"
# The protocol to be used for communicating with the target.
protocol = "Swd"
# The speed in kHz of the data link to the target.
speed = 100
```

Here's the full `Embed.toml` for the project (without the speed change):
https://cgit.pinealservo.com/BluePill_Rust/blue_pill_base/src/branch/master/Embed.toml

Any idea why `cargo-flash` works from master while `cargo-embed` does not? That seems like a clue.

**Comment by Yatekii** [comment]
Both succeed to flash :) `cargo-embed` only fails to attach to RTT :) But without seeing the full message it's hard to tell :) Maybe it was just a fluke, maybe it's real trouble, I can't say :) Normally when flashing works RTT works too :)

Depending on your blue pill it might be a HW issue :)

**Comment by coolaj86** [comment]
@Yatekii 

> without seeing the full message it's hard to tell

How can I print the full message? I don't see a `--verbose` option to `cargo embed`.

> Depending on your blue pill it might be a HW issue.

Why would it work with an older version just fine if it were hardware?

Luckily, I have a box full of these. I just tried two more. All 3 failed.

**Comment by Yatekii** [comment]
> Why would it work with an older version just fine if it were hardware?

Good point! So RTT works withg older versions? Sorry I didn't account for that :/

> How can I print the full message? I don't see a `--verbose` option to `cargo embed`.

I think you omitted parts of the message stack printed, am I wrong? You can set the log level in the Embed.toml to DEBUG instead of WARN :)

Hmm it's odd that flashing would work but RTT fails :/ I guess the only way to really debug this might be bisecting commits :/ I am kinda poking into the dark here ... real hard to tell remotely :/

**Comment by coolaj86** [comment]
Hmmm... unfortunately there's no more error output:

```txt
    Finished flashing in 4.226s
        INFO cargo_embed                            > Initializing RTT (attempt 1)...
       DEBUG probe_rs_rtt::rtt                      > Scanning at exact address: 0x20000000
       DEBUG probe_rs::probe::stlink                > Read mem 8 bit, address=20000000, length=24
       DEBUG probe_rs::probe::stlink                > Read mem 8 bit, address=20000018, length=24
       DEBUG probe_rs::probe::stlink                > Read mem 8 bit, address=0800177b, length=128
       DEBUG cargo_embed                            > Failed to initialize RTT. Retrying until timeout.
        INFO cargo_embed                            > Initializing RTT (attempt 2)...
       DEBUG probe_rs_rtt::rtt                      > Scanning at exact address: 0x20000000
       DEBUG probe_rs::probe::stlink                > Read mem 8 bit, address=20000000, length=24
       DEBUG probe_rs::probe::stlink                > Read mem 8 bit, address=20000018, length=24
       DEBUG probe_rs::probe::stlink                > Read mem 8 bit, address=0800177b, length=128
       DEBUG cargo_embed                            > Failed to initialize RTT. Retrying until timeout.
        INFO cargo_embed                            > Initializing RTT (attempt 3)...
       DEBUG probe_rs_rtt::rtt                      > Scanning at exact address: 0x20000000
       DEBUG probe_rs::probe::stlink                > Read mem 8 bit, address=20000000, length=24
       DEBUG probe_rs::probe::stlink                > Read mem 8 bit, address=20000018, length=24
       DEBUG probe_rs::probe::stlink                > Read mem 8 bit, address=0800177b, length=128
       DEBUG cargo_embed                            > Failed to initialize RTT. Retrying until timeout.
       DEBUG probe_rs::probe::stlink                > Read mem 32 bit, address=e0002000, length=1
       DEBUG probe_rs::probe::stlink                > Current device mode: Jtag
       Error Error attaching to RTT: Error communicating with probe: An error with the usage of the probe occured
```

> real hard to tell remotely

If you'd care to give me an SSH key, I'll be happy to let you in to <stm32@stm32.duckdns.org>. :)

**Comment by Yatekii** [comment]
Sure :) Maybe not tonight anymore tho :D Maybe you can also join the Matrix channel for easier comms :)

**Comment by coolaj86** [comment]
@Yatekii Could you cut a new release?

The latest master is working again:

```bash
cargo install --force --git https://github.com/probe-rs/cargo-embed
```

I can now see debug messages in the terminal. :D

**Comment by Yatekii** [comment]
Yeah, I think we should be able to make a release :)

**Comment by coolaj86** [comment]
@Yatekii ETA on the release?

**Comment by Yatekii** [comment]
I don't know. We are still battling some issues.

**Comment by Yatekii** [comment]
Afaik this issue is resolved and the fix will be part of a iminent 0.10.0 release. Please reopen if this is still an issue :)
